Implement "group members assume state" option for ZHA

TheJulianJES opened the pull request:
## Proposed change
### Summary
- makes light group members assume state instantly when the group is changed (like other integrations do: Hue/Z2M)
- makes light groups and light group members properly reflect what their actual (target) state is (especially during transitions)
- makes light group state changes much faster and more fluid (videos below)
- (adds a config option for doing all of the above -- currently disabled by default, would consider the opposite though)
- fixes an issue where group state can get stuck if a member is changed during a (long) transition
- fixes a bunch of issues with off_brightness and correctly populates it to the members (see [this commit](https://github.com/home-assistant/core/pull/84938/commits/991331673bd03bb99f07c6ae57367a5cc773c7b0) for details)
- improves (large) light groups a lot for slower Zigbee networks or slow coordinators (Conbee)

### Long version

This adds a config option called "Group members assume state of group" (disabled by default).
When enabled, this config option makes all group members of a ZHA group assume the state of that group (after successful broadcast calls).
The reason this is beneficial is that long transition times for ``light.turn_on`` with ZHA groups causes all group members to be left in the old state for the transition time. If a 1 second transition time is used, it can be seen as "lag" in the UI. Longer transition times are a bit worse though.

So I looked at how the Hue integration and Z2M are dealing with this.
And they always assume state of the group (there are some issues with the Hue integration (or Bridge?) where light levels are incorrectly updated), but at least for Z2M, I've even disconnected all light groups members from mains and they still showed their group color in HA (so the state is always assumed for all members there).

There's also not really a better way, since ZCL doesn't provide a way to get the target color, target color temp, nor target level. Simply also polling at the start of a longer transition is counter-productive and won't work, as we'd set it to the wrong colors, temp, and possibly brightness.

Like with the option disabled, all light bulbs are still polled at the end of the transition (so when they have target level/color/temp).

All in all, this config option will deliver:
- a faster and more responsive UI when using ZHA groups (especially when the default 0s transition time is changed to 1s -> 1.25s delay + polling)
- an experience closer to the one of other Zigbee integrations (Hue, Z2M, possibly deCONZ?)
- fixes a bunch of issues (especially when using transitions and "off brightness" stuff with groups)

The only downside is that group members could turn on in the UI, even when they're unplugged (but not yet marked as unavailable by ZHA). However eventually, the bulbs will either get polled (end of transition and then in a random interval), or they'll turn unavailable.
And it's apparently also not an issue with the Hue integration or Z2M (same thing happens, but no complaints).
But because of this, I set the config option to be disabled by default for now (although the user experience is improved much with little downside IMO)

### Bugfix
Even with the config option disabled, this PR fixes an issue where the group state would not update correctly (commit https://github.com/home-assistant/core/pull/84938/commits/ecf2cb85afc4f727a0400d9f4675460808364eff).
This code wasn't needed with the final implementation of the "brightness slider transition fix", as groups set their members to "transitioning" mode where they won't update at all.
(The bug fix is also required by the "group members assume state" part)

Removing this check in the LightGroup fixes multiple issues. Previously, the state of a group was completely wrong when doing the following:
- turn on group with 10 second transition
- turn on members individually
- turn off members individually
- group state would not update correctly (re-calculation was blocked)

With the config option enabled, many more issues with inconsistent state between groups and members are also fixed.
It also correctly populates "off brightness" down to group members (which is not possible without the config option), so a lot of issues with transitions are fixed too by always having the correct state on both the group and members.

### Comparison

NEW BEHAVIOR BELOW (config option enabled) (default transition time doesn't matter now):

https://user-images.githubusercontent.com/6409465/210189412-d93355a4-3b0f-42af-b5f2-6c861c1cfdef.mov

___

OLD BEHAVIOR BELOW (config option disabled) (using default 0s transition. with 1s, it's worse):

https://user-images.githubusercontent.com/6409465/210189418-473d8d04-ef7e-4c43-a940-6f9ccf2a0932.mov

---

At this point, I'd even consider enabling the config option by default, as it provides a much better user experience (IMO) (especially when using longer transition times with service calls (<- was **not** even shown in the video)) and the Hue integration and Z2M also assume member state. (deCONZ possibly too -- haven't tried though)

Note: For ZHA switch groups (and other ZHA group types), the behavior remains completely unchanged.
Here, I don't think it makes sense to assume the state of group members instantly, as there is no "transition waiting" to do.

### Further stuff
Whilst working on this, I noticed that deactivating the "color loop" effect (which is done by using ``light.turn_on`` with anything but an effect) is somewhat broken when providing it with a color, as it will briefly change the color of the color loop and only then end the color loop but all lamps then switch to their starting color. (So the color in the service call is basically ignored)
I don't really want to clutter this PR with color loop fixes (and no one uses it anyway) but I'll likely follow up on this in a future PR.

## Type of change

(I guess it's mostly a new feature, but it could also be considered as being a bugfix probably.)

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

TheJulianJES committed:
Initial "group members assume state" implementation for ZHA

TheJulianJES committed:
Remove left-over debug flag (where polling was disabled)

TheJulianJES committed:
Implement _send_member_assume_state_event() method and also use after turn_off

TheJulianJES committed:
Only assume updated arguments from service call to group

TheJulianJES committed:
Make code more readable and change checks slightly

TheJulianJES committed:
Move "send member assume state" events to LightGroup on/off calls

TheJulianJES committed:
Include new config option in tests

TheJulianJES committed:
Check that member is available before updating to assumed state

TheJulianJES committed:
Lower "update group from child delay" for debouncer to basically 0 when using assumed member state

TheJulianJES committed:
Allow "child to group" updates regardless of config option

This is not needed, as group members will not update their state, as long as they're transitioning. (If a group transitions, it also sets its members to transitioning mode)

This fixes multiple issues. Previously, the state of a group was completely wrong when:
- turn on group with 10 second transition
- turn on members individually
- turn off members individually
- group state would not update correctly

TheJulianJES committed:
Move "default update group from child delay" constant

TheJulianJES committed:
Change to new constant name in test

TheJulianJES committed:
Also update fan test to new constant name

TheJulianJES committed:
Decrease "update group from child delay" to 10ms

In my testing, 0.0 also works without any issues and correctly de-bounces child updates when using the "assume state option".
This is just for avoiding multiple state changes when changing the group -> children issue individual updates.
With 2 children in a group and delay 0, both child updates only cause one group re-calculation and state change.

0.01 (10ms) should be plenty for very slow systems to de-bounce the update (and in the worst case, it'll cause just another state change but nothing breaks)

TheJulianJES committed:
Also implement "assuming state" for effect

Not sure if anybody even uses this, but this one is a bit special because the effect is always deactivated if it's not provided in the light.turn_on call.

TheJulianJES commented:
This should be ready for a first review now.
I still want to test this on my main network, but it worked very well in my testing so far.

TheJulianJES committed:
Move shortened delay for "assuming members" to a constant

TheJulianJES committed:
Add basic test to verify that group members assume on/off state

TheJulianJES committed:
Move _assume_group_state function declaration out of async_added_to_hass

TheJulianJES committed:
Fix rare edge-case when rapidly toggling lights and light groups at the same time

This prevents an issue where either the group transition would unset the transition flag or the single light would unset the group transition status midst-transition.

Note: When a new individual transition is started, we want to unset the group flag, as we actually cancel that transition.

TheJulianJES committed:
Check that effect list exists, add return type

TheJulianJES committed:
Re-trigger CI due to timeout

TheJulianJES committed:
Increase ASSUME_UPDATE_GROUP_FROM_CHILD_DELAY slightly

The debouncer is used when updating group member states either by assuming them (in which case we want to barely have any delay), or between the time we get the results back from polling (where we want a slightly longer time).
As it's not easily possible to distinguish if a group member was updated via assuming the state of the group or by the polling that follows, 50 ms seems to be a good middle point.

TheJulianJES committed:
Add debug print for when updating group state

TheJulianJES committed:
Fix issues with "off brightness" when switching between group/members

This fixes a bunch of issues with "off brightness" and passes it down to the members correctly.
For example, if a light group is turned off with a transition (so bulbs get their level set to 1), this will also set the "off brightness" of all individual bulbs to the last level that they were at.

(It really fixes a lot of issues when using the "member assume group state" option. It's not really possible to fix them without that.)

Furthermore, issues where polling was previously needed to get the correct state after "playing with transitions", should now get be resolved and get correct state when using the "members assume group state" option.

Note: The only case which still can't be fixed is the following:
If individual lights have off_with_transition set, but not the group, and the group is then turned on without a level, individual lights might fall back to brightness level 1 (<- at least now shows correctly in UI even before polling).
Since all lights might need different brightness levels to be turned on, we can't use one group call. But making individual calls when turning on a ZHA group would cause a lot of traffic and thereby be counter-productive.
In this case, light.turn_on should just be called with a level (or individual calls to the lights should be made).

Another thing that was changed is to reset off_with_transition/off_brightness for a LightGroup when a member is turned on (even if the LightGroup wasn't turned on using its turn_on method).
off_with_transition/off_brightness for individual bulbs is now also turned off when a light is detected to be on during polling.

Lastly, the waiting for polled attributes could previously cause "invalid state" to be set (so mid-transition levels).
This could happen when group and members are repeatedly toggled at similar times. These "invalid states" could cause wrong "off brightness" levels if transitions are also used.
To fix this, we check after waiting for the polled attributes in async_get_state to see if a transition has started in the meanwhile. If so, the values can be discarded. A new poll will happen later and if using the "members assume group state" config option, the values should already be correct before the polling.

TheJulianJES commented:
After another day with messing with Zigbee lights, I've fixed quite a few issues with transitions + off_brightness + groups and individual lights in this commit: https://github.com/home-assistant/core/pull/84938/commits/991331673bd03bb99f07c6ae57367a5cc773c7b0
I've made some comments to make clear why something is done and also added a description for the above commit.

Of course, I'll do more testing, but this *should* now actually be ready for a first/second review.

Just to clarify: the behavior without activating the new config option is basically unchanged.
However, after it's activated, the behavior improves *a lot*. UI responsiveness is the biggest improvement probably. 
And using service calls with transitions (and switching that between groups and members) is also improved and fixes a bunch of issues with the current implementation.

lawrencedudley commented:
I'm a huge fan of this, it causes me a load of pain at the moment

puddly reviewed (approved):
Awesome!

I've been testing this on my own network and everything works as expected. I think this will be a good default.

TheJulianJES committed:
Enable "group members assume state" config option by default

The config tests are also updated to expect the config option be enabled by default.

For all tests, the config option is generally disabled though:
There are only two group related tests. The one that tests this new feature overrides the config option to be enabled anyway.
The other tests works in a similar way but also "sends" attribute reports, so we want to disable the feature for that test.
(It would also run with it enabled (if the correct CHILD_UPDATE value is patched), but then it would test the same stuff as the other test, hence we're disabling the config option for that test.)

TheJulianJES commented:
Enabled the config option by default now (and slightly updated the tests, so the one group test still tests without the feature and the other one with it).

If it's decided to keep the option disabled by default, the last commit can just be reverted.

TheJulianJES commented:
@puddly Is this still good and fine for the next beta/merging?
